[Keith Tan] iP

nathtiekk opened the pull request:
# Kif

> Kif frees your mind of having to remember things you need to do.

It is:
- text-based
- fast to use
- ~easy~ SUPER EASY to learn

All you need to do is:
1. download it from [here.](https://github.com/nathtiekk/ip/releases/tag/A-Gradle)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😃 

Features:
- [X] Managing tasks
- [ ] Managing deadlines (_coming soon_)
- [ ] Reminders (_coming soon_)

> [!TIP]
> If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

nathtiekk committed:
Level 0. Rename, Greet, Exit

nathtiekk committed:
Level 1. Echo

nathtiekk committed:
Level 2. Add, List

nathtiekk committed:
Level 3. Mark as Done

nathtiekk committed:
Level 4. ToDos, Events, Deadlines

nathtiekk committed:
A-TextUiTesting

nathtiekk committed:
Level 5. Handle Errors

nathtiekk committed:
Level 6. Delete

changjy81 commented on src/main/java/Kif.java:
```diff
@@ -0,0 +1,74 @@
+
+    private static void introduce() {
+        String introduce =
+                """
```
Huh, I wasn't aware of text blocks in Java. Good idea to use them!
Though I'm not sure of the code style requirements for their indentation...

changjy81 commented on src/main/java/Kif.java:
```diff
@@ -0,0 +1,74 @@
+        String[] splitMessage = userMessage.split(" ");
+        while(!isTerminate(splitMessage[0])) {
+            UserCommand command = UserCommand.valueOf(splitMessage[0].toUpperCase());
+            switch (command) {
+                case LIST -> Task.listUserTask();
+                case MARK -> Task.markUserTask(Integer.parseInt(splitMessage[1]));
+                case UNMARK -> Task.unmarkUserTask(Integer.parseInt(splitMessage[1]));
+                case DEADLINE -> Task.Deadline.createDeadline(userMessage);
+                case EVENT -> Task.Event.createEvent(userMessage);
+                case DELETE -> Task.delete(Integer.parseInt(splitMessage[1]));
+                case TODO -> {
+                    try {
+                        Task.ToDo.createToDo(userMessage);
+                    } catch (KifException e) {
+                        System.out.println(e.getMessage());
+                    }
+                }
+                default -> System.out.println(
+                                    """
+                                    ____________________________________________________________
+                                    OOPS!!! I'm sorry, but I don't know what that means :-(
+                                    ____________________________________________________________"""
+                );
+            }
```
The case statements shouldn't be indented according to the coding standard?

changjy81 commented on src/main/java/Kif.java:
```diff
@@ -0,0 +1,74 @@
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+
```
I see that the week 3 iP tasks were not pushed yet, but the classes should be in packages and should have header Javadoc comments on classes/methods

changjy81 replied:
I see that the week 3 iP tasks were not pushed yet, but the classes should be in packages and should have header Javadoc comments on classes/methods

changjy81 commented on src/main/java/Kif.java:
```diff
@@ -0,0 +1,74 @@
+        introduce();
+        String userMessage = reader.readLine();
+        String[] splitMessage = userMessage.split(" ");
+        while(!isTerminate(splitMessage[0])) {
```
Shouldn't there be a whitespace after while here?
Noticed similar occurences in the rest of the code as well.

changjy81 replied:
Shouldn't there be a whitespace after while here?
Noticed similar occurences in the rest of the code as well.

changjy81 reviewed (changes requested):
put wrong review type; redone below

changjy81 commented on src/main/java/Kif.java:
```diff
@@ -0,0 +1,74 @@
+
+    private static void introduce() {
+        String introduce =
+                """
+                ____________________________________________________________
+                Hello! I'm Kif
+                What can I do for you?
+                ____________________________________________________________
+                """;
```
Huh, I wasn't aware of text blocks in Java. Good idea to use them!
Though I'm not sure of the code style requirements for their indentation...

changjy81 commented on src/main/java/Kif.java:
```diff
@@ -0,0 +1,74 @@
+        String[] splitMessage = userMessage.split(" ");
+        while(!isTerminate(splitMessage[0])) {
+            UserCommand command = UserCommand.valueOf(splitMessage[0].toUpperCase());
+            switch (command) {
+                case LIST -> Task.listUserTask();
+                case MARK -> Task.markUserTask(Integer.parseInt(splitMessage[1]));
+                case UNMARK -> Task.unmarkUserTask(Integer.parseInt(splitMessage[1]));
+                case DEADLINE -> Task.Deadline.createDeadline(userMessage);
+                case EVENT -> Task.Event.createEvent(userMessage);
+                case DELETE -> Task.delete(Integer.parseInt(splitMessage[1]));
+                case TODO -> {
+                    try {
+                        Task.ToDo.createToDo(userMessage);
+                    } catch (KifException e) {
+                        System.out.println(e.getMessage());
+                    }
+                }
+                default -> System.out.println(
+                                    """
+                                    ____________________________________________________________
+                                    OOPS!!! I'm sorry, but I don't know what that means :-(
+                                    ____________________________________________________________"""
+                );
```
The case statements shouldn't be indented according to the coding standard?

changjy81 reviewed (commented):
Outside of some recurring coding standard violations, LGTM overall.

nathtiekk committed:
branch-Level-7

nathtiekk committed:
Merge branch 'branch-Level-7'

gohlucas commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,271 @@
+    protected boolean isDone;
+    private static final String FILEPATH = "tasks.txt";
+    protected taskType type;
+    private static final String KEYWORD = "kifReservedKeyword";
```
Would grouping variables by their access modifiers be better? For instance, grouping the private static final variables together etc...

gohlucas commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,271 @@
+                }
+            }
+        } catch (FileNotFoundException e) {
+            //do nothing
```
I would recommend handling the exception. This could be in the form of printing an error message to the user to inform them that there was an issue with the ongoing process

gohlucas commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,271 @@
+    private static void editTaskTXT (int lineNumber, Kif.UserCommand operation) {
+        ArrayList<String> preItems = new ArrayList<>();
+        ArrayList<String> postItems = new ArrayList<>();
+        ArrayList<String> editLine = new ArrayList<>();
```
Would changing the variable name to a plural form be more inline with the Java Coding Standard for collection of objects?

gohlucas commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,271 @@
+    }
+
+    public static void listUserTask() {
+        System.out.println("____________________________________________________________");
```
Would extracting the line into a static final variable provide a cleaner body for the methods? I noticed this line is used quite extensively in this file

gohlucas commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,271 @@
+    public static void markUserTask(int key) {
+        Task currentTask = userTasks.get(key - 1);
+        currentTask.isDone = true;
+        editTaskTXT(key, Kif.UserCommand.MARK);
```
```suggestion
        editTaskTxt(key, Kif.UserCommand.MARK);
```

gohlucas replied:
Perhaps this would be more in line with the java coding standard

gohlucas commented on src/main/java/Kif.java:
```diff
@@ -0,0 +1,80 @@
+        Task.listUserTask();
+    }
+
+    public enum UserCommand {
+        LIST,
+        MARK,
+        UNMARK,
+        DEADLINE,
+        EVENT,
+        TODO,
+        DELETE;
+    }
```
I love that you created an enum for the types of commands!!

gohlucas reviewed (commented):
I can see the vision. Looking forward to Kif's development 👍

nathtiekk committed:
Level-8

nathtiekk committed:
Merge branch 'branch-Level-8'

nathtiekk committed:
A-Packages

nathtiekk committed:
Merge branch 'add-gradle-support'

nathtiekk committed:
A-Gradle

nathtiekk committed:
A-Jar

nathtiekk committed:
A-CodingStandard

nathtiekk committed:
branch-A-CodingStandard

nathtiekk committed:
Merge branch 'branch-A-CodingStandard'

nathtiekk committed:
A-JavaDoc

nathtiekk committed:
Merge branch 'branch-A-JavaDoc'

nathtiekk committed:
A-MoreOOP

nathtiekk committed:
Integrate with JavaFX tutorial

nathtiekk committed:
Add Level-10

nathtiekk committed:
Merge branch 'branch-Level-10'

nathtiekk committed:
BCD-Extension: Undo

nathtiekk committed:
Merge branch 'branch-BCD-Extension'

nathtiekk committed:
Refactor handleCommand method

The handleCommand method has been refactored to improve its structure, making the code more readable and maintainable.\nError handling has been improved for commands that require parsing user input (e.g., MARK, UNMARK, DELETE). This improves the robustness of the method by preventing runtime errors due to invalid user input.\nAlso added a default case in the switch statement to catch any unknown commands, ensuring that unexpected user input is properly flagged with an appropriate error message.\nChanges include:\n- Refactored previousCommand and previousTask handling.\n- Improved error message for unknown commands.

nathtiekk committed:
Merge pull request #2 from nathtiekk/branch-A-CodeQuality

Refactor handleCommand method

nathtiekk committed:
Add A-UserGuide

nathtiekk committed:
Update README.md

nathtiekk committed:
Add A-Assertion

nathtiekk committed:
Merge branch 'branch-A-Assertions'

nathtiekk committed:
Add A-JUnit

nathtiekk committed:
Add more JavaDoc
